### DISCUSSION TITLE
Scala 3 extractors and Scala 2 binary compatibility

### DISCUSSION THREAD

**Discussion by kincsescsaba** [scala/scala3#21563] [Compiler internals Q&A] [answered]
While examining a library-specific binary compatibility issue - which can be seen with its error messages here: [https://github.com/scala/scala3/issues/21560](https://github.com/scala/scala3/issues/21560) - I found that it was caused by a more general phenomenon, namely that `unapply` methods of case classes apparently have different return types in Scala 2 and Scala 3.

I created a demo project to illustrate this: [https://github.com/kincsescsaba/scala-compiler-crash-cc-unapply/tree/master](https://github.com/kincsescsaba/scala-compiler-crash-cc-unapply/tree/master)

If I run the demo, I get this output:
```
Scala 2:
public scala.Option io.github.kincsescsaba.libcompiledasscala2.Scala2CC$.unapply(io.github.kincsescsaba.libcompiledasscala2.Scala2CC)
Scala 3:
public io.github.kincsescsaba.crossscala3scala2.Main$Scala3CC io.github.kincsescsaba.crossscala3scala2.Main$Scala3CC$.unapply(io.github.kincsescsaba.crossscala3scala2.Main$Scala3CC)
```

Looks like a Scala 2 case class `unapply` method have the return type `Option[CC]` while a Scala 3 `unapply` has `CC`.

If you check the library-specific issue cited above, you can see that the runtime looked for a Scala 3 specific `unapply` method, while a Scala 2 specific was available.

Even though Scala 2 and 3 should be binary compatible, looks like this issue was not handled.

I'd like to understand if the different return type is a bug or a feature, is this issue common and is there a workaround in plan? Like generating/accepting both `unapply` signatures.
Most likely this can cause crashes in different libraries, hence I shared the details of it.

**Comment by dwijnand** [comment, marked as answer]
So Scala 2 is binary compatible with Scala 3, in that the Scala 3 compiler can understand and compile code against a Scala 2 library.  But Scala 3 decided to change the return type of case class unapplies, and that _is_ an improvement in Scala 3.  But it's still able to consider Scala 2's unapplies as extractors, and case class extractors at that (i.e. irrefutable).  If you use reflection, you'll see the difference - but that doesn't make them "incompatible".

**Reply by kincsescsaba** [reply]
Thanks for looking into this! I accept that, I just state that there are possibly edge cases when this still caused a crash and like to figure out what is the best solution to handle this. Including learning about if a fix is planned or not

**Comment by som-snytt** [comment]
This question was asked in another forum, but I can't find it. What is the link?

**Reply by kincsescsaba** [reply]
I asked it in a different form when I thought this is ScalaJs specific. When I saw it is more general, I opened this discussion with details highlighting what can be useful to examine it from that side.
I think you saw this issue: [https://users.scala-lang.org/t/adding-scalajs-linker-as-a-dependency-introduces-unexpected-error/10205](https://users.scala-lang.org/t/adding-scalajs-linker-as-a-dependency-introduces-unexpected-error/10205)
